token-lending/program/src/state.rs

```diff
@@ -525,3 +596,87 @@ fn pack_decimal(decimal: Decimal, dst: &mut [u8; 16]) {
 fn unpack_decimal(src: &[u8; 16]) -> Decimal {
     Decimal::from_scaled_val(u128::from_le_bytes(*src))
 }
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::math::WAD;
+    use proptest::prelude::*;
+
+    proptest! {
+        #[test]
+        fn borrow_fee_calculation(
+            borrow_fee_wad in 0..WAD,
+            host_fee_percentage in 0..100u8,
+            borrow_amount in 0..u64::MAX,
+        ) {
+            let fees = ReserveFees {
+                borrow_fee_wad,
+                host_fee_percentage,
+            };
+            let (total_fee, host_fee) = fees.calculate_borrow_fees(
+                borrow_amount,
+                &Pubkey::new_unique(),
+            );
+
+            // the total fee can't be greater than the amount borrowed
+            assert!(total_fee <= borrow_amount);
+            // the host fee can't be greater than the total fee
+            assert!(host_fee <= total_fee);
+
+            // for all fee rates greater than 0, we must have some fee
+            if borrow_fee_wad > 0 {
+                assert!(total_fee > 0);
+            }
+
+            let host_fee_percentage = host_fee_percentage as u64;
```

Comment:
nit: I don't think this conversion is needed

Reply:
Nice catch